### DISCUSSION TITLE
Using pydantic / pydantic settings and supporting Python 3.6

### DISCUSSION THREAD

**Discussion by maartenbreddels** [pydantic/pydantic#6497] [Question] [answered]
Hi,

I loved using Pydantic for settings support and used it in https://github.com/vaexio/vaex/ and https://github.com/widgetti/solara/ 
Both libraries try to provide Python 3.6 support, since large enterprises still use that and for various reasons do not upgrade.
With the release of pydantic 2.0, the settings support is split off in pydantic-settings.
Ideally, I do not want to pin pydantic < 2 in any of my packages, nor do I want to drop Python 3.6 support, however this seems not possible because if I require `"pydantic", "pydantic-settings; python_version > '3.6'",` as dependency, I effectively require pydantic 2 for Python > 3.6? 

Is it possible to be compatible with pydantic 1 and 2 for Python >= 3.6?

Regards,

Maarten

**Comment by hramezani** [comment, marked as answer]
Hi @maartenbreddels 

I think it is not possible to add Python 3.6 to Pydantic V2. Even [Pydantic V1 doesn't support Python 3.6 right now](https://github.com/pydantic/pydantic/blob/15c82d978d9e3b3893f00982b9e0c5c8d5c7821d/setup.py#L131).

**Comment by Kludex** [comment]
On a note, I don't think any Python package should take the decision of supporting a Python version that is in EOL. We actually should make efforts to make sure that people don't use Python versions that are not receiving security updates anymore.

**Reply by itkovian** [reply]
Meaning people using RHEL whatever cannot get the latest python packages, as RHEL is stuck with whatever they chose at some point.

**Comment by maartenbreddels** [comment]
> I think it is not possible to add Python 3.6 to Pydantic V2

I am not asking for that, I'm still happy to use an older pydantic version for when people want to use my libraries with Python 3.6 However, I don't want to force my user into a specific version (pydantic v2) when not needed.

@Kludex in practice we see users in enterprises (who are indirectly paying for many developers bills) not being able to upgrade. They may not have permissions to do so, or because of incompatibilities now be able to upgrade.

I think what happened is that many companies switch over to 3, when 3.6 is out, and we're left with that situation. In many cases security is not an issue. If you run a Jupyter notebook, or batch script, all bets are off anyway, a pip install could even rm your whole home directory. If you are talking about a server, you are right, security is important, but this is just what we see in practice.

I'm afraid that this situation will lead to upper bound pinning (e.g. pydantic < 2) which will hurt users even more. EOL does not mean end of use.

Maybe a solution would be to have `pydantic[settings]` as explained here: https://github.com/pydantic/pydantic/pull/4492

It would be great if libraries that use pydantic + settings do not have to choose v1 or v2, but be compatible with both. That is currently not possible.

PS: Note that there could also be different reasons to stay with pydantic 1 (e.g. a bug in 2), or a change in behavior.

**Comment by maartenbreddels** [comment]
A bit off-topic, but just to illustrate the pains of upper bound pins.

The previous owner of pydantic-settings used poetry (which should not be used for libraries I think, just applications) to put an unneeded upper bound on `typing-extensions` https://github.com/danields761/pydantic-settings/blob/bfdb531526306130c6dd0a704d197b18e2b0e72a/pyproject.toml#L29C1-L29C29 which closes off another escape hatch out of this situation which interferes with a (valid and needed!) lower limit we have on `typing-extensions`. If this wasn't the case, we could just have mixed `pydantic-settings` with `pydantic<2`.

**Comment by maartenbreddels** [comment]
It seems the pydantic.v1 package could have saved us (see https://github.com/pydantic/pydantic/discussions/4649) but before we found, we already took pydantic out https://github.com/widgetti/solara/pull/189

**Comment by samuelcolvin** [comment]
Hi @maartenbreddels, I'm not really sure what you were asking for?

If you're requesting that the `pydantic-settings` package supports pydantic v1.X, I really don't think that's practical at this point.

Sorry for your frustrations, but I agree with @Kludex that supporting 3.6 as this point is a mistake. Even 3.7 is about to go end of life.

I can understand those who want to keep existing code running on 3.6, but I have no idea why anyone (in any size of company) would want to be doing active development with 3.6?

Having dropped 3.6 from Pydantic over a year ago, AFAIK this is the first request we've had for 3.6 support (maybe there are a few others, but really not many), so I don't think your case is very common.

**Comment by maartenbreddels** [comment]
I think I should not have brought up Python 3.6, since the issue is more general than that.

The main problem occurs when a library is using the BaseSettings feature. By splitting pydantic for version 2 into pydantic and pydantic-setting it makes it impossible for libraries to support pydantic v1 *and* v2. Applications can easily choose a version, and pin it. But if libraries want to be cooperative, they should try to support both (and not pin). 

Using the dependency system in Python I cannot have `pydantic-settings` be an optional dependency, where the condition should be something like `pydantic >= 2.0.0`.

If I understand correctly, I should have used the `pydantic.v1` package for backward compatibility? If so, I highly recommend putting this into the migration docs and in the deprecation warning. I can imagine libraries resorting to pinning (v1 or v2), leading to unsolvable environments.

**Reply by Pkoiralap** [reply]
I am facing this. I am using two libraries that use pydantic 1 and pydantic 2. And I can't seem to find a version that both those libraries agree on. `pydantic-settings` must be allowed to work with pydantic 1
